internal/util/importutil/csv_parser.go

```diff
@@ -0,0 +1,306 @@
+		bs, err := json.Marshal(mp)
+		if err != nil {
+			log.Warn("CSV parser: illegal value for dynamic field, not a JSON object")
+			return errors.New("illegal value for dynamic field, not a JSON object")
+		}
+		row[dynamicFieldID] = string(bs)
+	} else if !ok && len(dynamicValues) == 0 {
+		// case 5
+		row[dynamicFieldID] = "{}"
+	}
+	// else case 3
+
+	return nil
+}
+
+func (p *CSVParser) verifyRow(raw []string) (map[storage.FieldID]string, error) {
+	row := make(map[storage.FieldID]string)
+	dynamicValues := make(map[string]string)
+
+	for i := 0; i < len(p.fieldsName); i++ {
+		fieldName := p.fieldsName[i]
+		fieldID, ok := p.collectionInfo.Name2FieldID[fieldName]
+
+		if fieldID == p.collectionInfo.PrimaryKey.GetFieldID() && p.collectionInfo.PrimaryKey.GetAutoID() {
+			// primary key is auto-id, no need to provide
+			log.Warn("CSV parser: the primary key is auto-generated, no need to provide", zap.String("fieldName", fieldName))
+			return nil, fmt.Errorf("the primary key '%s' is auto-generated, no need to provide", fieldName)
+		}
+
+		if ok {
+			row[fieldID] = raw[i]
+		} else if p.collectionInfo.DynamicField != nil {
+			// collection have dynamic field. put it to dynamicValues
+			dynamicValues[fieldName] = raw[i]
+		} else {
+			// no dynamic field. if user provided redundant field, return error
+			log.Warn("CSV parser: the field is not defined in collection schema", zap.String("fieldName", fieldName))
+			return nil, fmt.Errorf("the field '%s' is not defined in collection schema", fieldName)
+		}
+	}
+	// some fields not provided?
+	if len(row) != len(p.collectionInfo.Name2FieldID) {
+		for k, v := range p.collectionInfo.Name2FieldID {
+			if p.collectionInfo.DynamicField != nil && v == p.collectionInfo.DynamicField.GetFieldID() {
+				// ignore dyanmic field, user don't have to provide values for dynamic field
+				continue
+			}
+
+			if v == p.collectionInfo.PrimaryKey.GetFieldID() && p.collectionInfo.PrimaryKey.GetAutoID() {
+				// ignore auto-generaed primary key
+				continue
+			}
+			_, ok := row[v]
+			if !ok {
+				// not auto-id primary key, no dynamic field,  must provide value
+				log.Warn("CSV parser: a field value is missed", zap.String("fieldName", k))
+				return nil, fmt.Errorf("value of field '%s' is missed", k)
+			}
+
+		}
+	}
+	// combine the redundant pairs into dynamic field(if has)
+	err := p.combineDynamicRow(dynamicValues, row)
+	if err != nil {
+		log.Warn("CSV parser: failed to combine dynamic values", zap.Error(err))
+		return nil, err
+	}
+
+	return row, nil
+}
+
+func (p *CSVParser) ParseRows(reader *IOReader, handle CSVRowHandler) error {
+	if reader == nil || handle == nil {
+		log.Warn("CSV Parser: CSV parse handle is nil")
+		return errors.New("CSV parse handle is nil")
+	}
+	r := csv.NewReader(reader.r)
+
+	oldPercent := int64(0)
+	updateProgress := func() {
+		if p.updateProgressFunc != nil && reader.fileSize > 0 {
+			percent := (r.InputOffset() * ProgressValueForPersist) / reader.fileSize
```

Comment:
csv `InputOffset()` is only supported after go 1.19. This will break milvus's go >=1.18 dependency restrictions. @xiaofan-luan Is that OK or we have to work around it?